pkg/finality-grandpa/dummy_chain_test.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/exp/slices"
+)
+
+const GenesisHash = "genesis"
+const nullHash = "NULL"
+
+type blockRecord struct {
+	hash   string
+	number uint32
+	parent string
+}
+
+// dummyChain is translation of finality_grandpa::testing::chain::DummyChain
+type dummyChain struct {
+	inner     map[string]blockRecord
+	leaves    []blockRecord
+	finalized struct {
+		hash   string
+		number uint32
+	}
+}
+
+func newDummyChain() *dummyChain {
+	dc := &dummyChain{
+		inner:  make(map[string]blockRecord),
+		leaves: make([]blockRecord, 0),
+	}
+	dc.inner[GenesisHash] = blockRecord{
+		number: 1,
+		parent: nullHash,
+		hash:   GenesisHash,
+	}
+	dc.leaves = append(dc.leaves, dc.inner[GenesisHash])
+	dc.finalized.hash = GenesisHash
+	dc.finalized.number = 1
+	return dc
+}
+
+func (dc *dummyChain) Ancestry(base, block string) (ancestors []string, err error) {
+	ancestors = make([]string, 0)
+loop:
+	for {
+		br, ok := dc.inner[block]
+		if !ok {
+			// TODO: make this sentinel error on entire package
```

Comment:
is this still a todo?

Reply:
no I don't think this is needed at the moment.